### PR TITLE
NIFI-13126: Add Impala DB Adapter

### PR DESCRIPTION
**Pull request by takraj**
# Summary

[NIFI-13126](https://issues.apache.org/jira/browse/NIFI-13126)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### UI Contributions

- [x] NiFi is modernizing its UI. Any contributions that update the [current UI](https://github.com/apache/nifi/tree/main/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui) also need to be implemented in the [new UI](https://github.com/apache/nifi/tree/main/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi).  

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/db/impl/TestImpalaDatabaseAdapter.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Test
+    public void testGetUpsertStatementWithEmptyKeyColumnNames() {
+        testGetUpsertStatement("notEmpty", Collections.singletonList("notEmpty"), Collections.emptyList(), new IllegalArgumentException("Key column names cannot be null or empty"));
+    }
+
+    @Test
+    public void testGetUpsertStatement() {
+        // GIVEN
```
Comment (dan-s1) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/db/impl/TestImpalaDatabaseAdapter.java:76]:
```suggestion
```

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/db/impl/TestImpalaDatabaseAdapter.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        String tableName = "table";
+        List<String> columnNames = Arrays.asList("column1", "column2", "column3", "column4");
+        Collection<String> uniqueKeyColumnNames = Arrays.asList("column2", "column4");
+
+        String expected = "UPSERT INTO" +
+                " table(column1, column2, column3, column4) VALUES (?, ?, ?, ?)";
+
+        // WHEN
+        // THEN
```
Comment (dan-s1) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/db/impl/TestImpalaDatabaseAdapter.java:85]:
```suggestion
```

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/db/impl/TestImpalaDatabaseAdapter.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    private void testGetUpsertStatement(String tableName, List<String> columnNames, Collection<String> uniqueKeyColumnNames, IllegalArgumentException expected) {
+        final IllegalArgumentException e = assertThrows(IllegalArgumentException.class, () -> {
+            testGetUpsertStatement(tableName, columnNames, uniqueKeyColumnNames, (String) null);
+        });
+        assertEquals(expected.getMessage(), e.getMessage());
+    }
+
+    private void testGetUpsertStatement(String tableName, List<String> columnNames, Collection<String> uniqueKeyColumnNames, String expected) {
+        // WHEN
```
Comment (dan-s1) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/db/impl/TestImpalaDatabaseAdapter.java:97]:
```suggestion
```

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/db/impl/TestImpalaDatabaseAdapter.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        String actual = testSubject.getUpsertStatement(tableName, columnNames, uniqueKeyColumnNames);
+
+        // THEN
```
Comment (dan-s1) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/test/java/org/apache/nifi/processors/standard/db/impl/TestImpalaDatabaseAdapter.java:100]:
```suggestion
```

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/db/impl/ImpalaDatabaseAdapter.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    public String getUpsertStatement(String table, List<String> columnNames, Collection<String> uniqueKeyColumnNames) {
+        if (StringUtils.isEmpty(table)) {
+            throw new IllegalArgumentException("Table name cannot be null or blank");
+        }
+        if (columnNames == null || columnNames.isEmpty()) {
+            throw new IllegalArgumentException("Column names cannot be null or empty");
+        }
+        if (uniqueKeyColumnNames == null || uniqueKeyColumnNames.isEmpty()) {
+            throw new IllegalArgumentException("Key column names cannot be null or empty");
+        }
```
Comment (dan-s1) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/db/impl/ImpalaDatabaseAdapter.java:67]:
```suggestion

        if (StringUtils.isEmpty(columnNames) {
            throw new IllegalArgumentException("Column names cannot be null or empty");
        }
        
        if (StringUtils.isEmpty(uniqueKeyColumnNames) {
            throw new IllegalArgumentException("Key column names cannot be null or empty");
        }
```

File: nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/db/impl/ImpalaDatabaseAdapter.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                .collect(Collectors.joining(", "));
+
+        String parameterizedInsertValues = columnNames.stream()
+                .map(__ -> "?")
+                .collect(Collectors.joining(", "));
+
+        List<String> updateValues = new ArrayList<>();
+        for (int i = 0; i < columnNames.size(); i++) {
+            updateValues.add(columnNames.get(i) + " = ?");
+        }
```
Comment (dan-s1) [inline nifi-extension-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/db/impl/ImpalaDatabaseAdapter.java:79]: Intellij is complaining with 

> Contents of collection 'updateValues' are updated, but never queried

**Comment by joewitt** [comment]
I am thinking we want to be careful about including this in the standard bundle.  Though I notice this doesn't have any apparent relationship to a particular binary/jdbc driver.  It does offer a SPI mechanism to find providers.  Does that mean we can have this in its own nar which could be optionally included?  My concern is how many people in the community will be able to participate/actively engage in maintaining this component.

**Comment by mattyb149** [comment]
Without Impala dependencies I don't see why it can't go into the standard NAR, especially since `PhoenixDatabaseAdapter` is a precedent. I think there are a number of people in the community (including myself) that can maintain this. Thoughts?

**Comment by takraj** [comment]
@mattyb149 , @joewitt 

How about doing some refactor *for 2.x*, and extracting these DB adapters as controller services? 

For 1.x line we could keep the current behavior for compatibility, and add an option to use controller service instead. Eg. the dropdown could have an option like 'Use Controller Service', and when that's selected the user can pick one.

Thoughts?

**Comment by mattyb149** [comment]
I didn't make them controller services at the time because they just didn't feel like controller services to me, they shouldn't need to be configured separately and just represent specific dialects of SQL. Having said that I'm not opposed to the idea if it makes sense going forward.

**Comment by takraj** [comment]
@mattyb149 @joewitt 

I have created a PoC / PR implementing the concept of having controller services instead of fixed lists of supported SQL dialects.

If we accept this solution, then Impala dialect is not necessarily needed to be included & maintained in the NiFi repository.

Please have a look:
https://github.com/apache/nifi/pull/8892

**Comment by exceptionfactory** [comment]
Thanks for putting forward the new Controller Service approach in #8892. I'm closing this pull request in favor of those changes, and will look to follow up there. If there turns out to be some blocker in the course of that review, this PR could be revisited.

**Pull request closed without merging**